src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java

```diff
@@ -449,6 +449,12 @@ public PlainPackageBinding getIncarnation(ModuleBinding arg0) {
 		CtExecutableReference<T> ref;
 		if (allocationExpression.binding != null) {
 			ref = getExecutableReference(allocationExpression.binding);
+			// in some cases the binding is not null but points wrong to object type see #4643
+			if (ref.getType().equals(ref.getFactory().Type().objectType()) && allocationExpression.resolvedType != null) {
+				ref.setType(getTypeReference(allocationExpression.resolvedType));
+				ref.getExecutableDeclaration().setType(getTypeReference(allocationExpression.resolvedType));
+				ref.setDeclaringType(getTypeReference(allocationExpression.resolvedType));
+			}
```

Comment:
I think we should make it explicit that this is a noclasspath workaround. In isolation like this, it's just another special case that will make no sense to anyone roughly two weeks from now.

I suggest something like this:

```suggestion
			if (isIncorrectlyBoundExecutableInNoClasspath(ref)) {
				adjustExecutableAccordingToResolvedType(allocationExpression, ref);
			}
```

And then have these definitions:

```java
private boolean isIncorrectlyBoundExecutableInNoClasspath(CtExecutableReference<?> ref) {
	boolean noClasspath = ref.getfactory().getEnvironment().getNoClasspath();
	return noClasspath
		&& ref.getType().equals(ref.getFactory().Type().objectType())
		&& allocationExpression.resolvedType != null;
}

private voied adjustExecutableAccordingToResolvedType(AllocationExpression allocationExpression, CtExecutableReference ref) {
	CtTypeReference<?> resolvedTypeRef = getTypeReference(allocationExpression.resolvedType);
	ref.setType(resolvedTypeRef);
	ref.getExecutableDeclaration().setType(resolvedTypeRef);
	ref.setDeclaringType(resolvedTypeRef);
}
```